File: arckit-plugin/hooks/arckit-context.py
#!/usr/bin/env python3
"""
ArcKit UserPromptSubmit Hook

Pre-computes project context when any /arckit: command is run.
Injects project inventory, artifact lists, and external documents
as a systemMessage so commands don't need to discover this themselves.

Hook Type: UserPromptSubmit
Input (stdin): JSON with user_prompt, cwd, etc.
Output (stdout): JSON with systemMessage containing project context
"""

import json
import os
import re
import sys
from pathlib import Path

# Doc type code to human-readable name mapping
DOC_TYPE_NAMES = {
    "PRIN": "Architecture Principles",
    "STKE": "Stakeholder Analysis",
    "REQ": "Requirements",
    "RISK": "Risk Register",
    "SOBC": "Business Case",
    "PLAN": "Project Plan",
    "ROAD": "Roadmap",
    "STRAT": "Architecture Strategy",
    "BKLG": "Product Backlog",
    "HLDR": "High-Level Design Review",
    "DLDR": "Detailed Design Review",
    "DATA": "Data Model",
    "WARD": "Wardley Map",
    "DIAG": "Architecture Diagram",
    "DFD": "Data Flow Diagram",
    "ADR": "Architecture Decision Record",
    "TRAC": "Traceability Matrix",
    "TCOP": "TCoP Assessment",
    "SECD": "Secure by Design",
    "SECD-MOD": "MOD Secure by Design",
    "AIPB": "AI Playbook Assessment",
    "ATRS": "ATRS Record",
    "DPIA": "Data Protection Impact Assessment",
    "JSP936": "JSP 936 Assessment",
    "SVCASS": "Service Assessment",
    "SNOW": "ServiceNow Design",
    "DEVOPS": "DevOps Strategy",
    "MLOPS": "MLOps Strategy",
    "FINOPS": "FinOps Strategy",
    "OPS": "Operational Readiness",
    "PLAT": "Platform Design",
    "SOW": "Statement of Work",
    "EVAL": "Evaluation Criteria",
    "DOS": "DOS Requirements",
    "GCLD": "G-Cloud Search",
    "GCLC": "G-Cloud Clarifications",
    "DMC": "Data Mesh Contract",
    "RSCH": "Research Findings",
    "AWRS": "AWS Research",
    "AZRS": "Azure Research",
    "GCRS": "GCP Research",
    "DSCT": "Data Source Discovery",
    "STORY": "Project Story",
    "ANAL": "Analysis Report",
    "PRIN-COMP": "Principles Compliance",
    "CONF": "Conformance Assessment",
}

# Multi-instance types that have sequence numbers in filenames
MULTI_INSTANCE_RE = re.compile(r"^([A-Z]+-?[A-Z]*)-\d{3}$")


def doc_type_name(code):
    """Get human-readable name for a doc type code."""
    return DOC_TYPE_NAMES.get(code, code)


def extract_doc_type(filename):
    """Extract doc type code from ARC filename.

    ARC-001-REQ-v1.0.md -> REQ
    ARC-001-ADR-001-v1.0.md -> ADR
    ARC-001-SECD-MOD-v1.0.md -> SECD-MOD
    """
    # Strip ARC-NNN- prefix
    m = re.match(r"^ARC-\d{3}-(.+)$", filename)
    if not m:
        return filename
    rest = m.group(1)
    # Remove version suffix: -vN.N.md or -vN.N
    rest = re.sub(r"-v\d+(\.\d+)?\.md$", "", rest)
    rest = re.sub(r"-v\d+(\.\d+)?$", "", rest)
    # Strip trailing -NNN for multi-instance types
    mm = MULTI_INSTANCE_RE.match(rest)
    if mm:
        return mm.group(1)
    return rest


def find_repo_root(cwd):
    """Find repo root by looking for projects/ directory."""
    current = Path(cwd).resolve()
    while current != current.parent:
        if (current / "projects").is_dir():
            return str(current)
        current = current.parent
    return None


def main():
    try:
        raw = sys.stdin.read()
        if not raw or not raw.strip():
            sys.exit(0)
        data = json.loads(raw)
    except (json.JSONDecodeError, ValueError, EOFError):
        sys.exit(0)

    user_prompt = data.get("user_prompt", "")

    # Only run for /arckit: commands
    if not user_prompt.startswith("/arckit:"):
        sys.exit(0)

    # Commands that don't need project context
    m = re.match(r"^/arckit:([a-z_-]*)", user_prompt)
    if m:
        command = m.group(1)
        if command in ("pages", "customize", "create", "init", "list", "trello"):
            sys.exit(0)

    # Find repo root
    cwd = data.get("cwd", os.getcwd())
    repo_root = find_repo_root(cwd)
    if not repo_root:
        sys.exit(0)

    projects_dir = os.path.join(repo_root, "projects")
    if not os.path.isdir(projects_dir):
        sys.exit(0)

    # Build context string
    lines = []
    lines.append("## ArcKit Project Context (auto-detected by hook)\n")
    lines.append(f"Repository: {repo_root}\n")

    # Count projects
    project_entries = sorted(
        e for e in os.listdir(projects_dir)
        if os.path.isdir(os.path.join(projects_dir, e))
    )
    lines.append(f"**{len(project_entries)} project(s) found:**\n")

    # Scan each project
    for project_name in project_entries:
        project_dir = os.path.join(projects_dir, project_name)

        # Extract project number
        project_number = ""
        pm = re.match(r"^(\d{3})-", project_name)
        if pm:
            project_number = pm.group(1)

        lines.append(f"### {project_name}")
        lines.append(f"- **Path**: {project_dir}")
        if project_number:
            lines.append(f"- **Project ID**: {project_number}")

        # Scan for ARC-* artifacts in main project dir
        artifact_list = []
        artifact_count = 0
        newest_artifact_mtime = 0

        for f in sorted(os.listdir(project_dir)):
            fp = os.path.join(project_dir, f)
            if os.path.isfile(fp) and f.startswith("ARC-") and f.endswith(".md"):
                dtype = extract_doc_type(f)
                dname = doc_type_name(dtype)
                artifact_list.append(f"  - `{f}` ({dname})")
                artifact_count += 1
                amtime = os.path.getmtime(fp)
                if amtime > newest_artifact_mtime:
                    newest_artifact_mtime = amtime

        # Also scan subdirectories
        for subdir in ("decisions", "diagrams", "wardley-maps", "data-contracts", "reviews"):
            sub_path = os.path.join(project_dir, subdir)
            if os.path.isdir(sub_path):
                for f in sorted(os.listdir(sub_path)):
                    fp = os.path.join(sub_path, f)
                    if os.path.isfile(fp) and f.startswith("ARC-") and f.endswith(".md"):
                        dtype = extract_doc_type(f)
                        dname = doc_type_name(dtype)
                        artifact_list.append(f"  - `{subdir}/{f}` ({dname})")
                        artifact_count += 1
                        amtime = os.path.getmtime(fp)
                        if amtime > newest_artifact_mtime:
                            newest_artifact_mtime = amtime

        if artifact_count > 0:
            lines.append(f"- **Artifacts** ({artifact_count}):")
            lines.extend(artifact_list)
        else:
            lines.append("- **Artifacts**: none")

        # Check for vendor directories
        vendors_dir = os.path.join(project_dir, "vendors")
        if os.path.isdir(vendors_dir):
            vendor_list = []
            for vname in sorted(os.listdir(vendors_dir)):
                vpath = os.path.join(vendors_dir, vname)
                if os.path.isdir(vpath):
                    vendor_list.append(f"  - {vname}")
            if vendor_list:
                lines.append(f"- **Vendors** ({len(vendor_list)}):")
                lines.extend(vendor_list)

        # Check for external documents
        external_dir = os.path.join(project_dir, "external")
        if os.path.isdir(external_dir):
            ext_list = []
            for f in sorted(os.listdir(external_dir)):
                fp = os.path.join(external_dir, f)
                if not os.path.isfile(fp):
                    continue
                if f == "README.md":
                    continue
                ext_mtime = os.path.getmtime(fp)
                if ext_mtime > newest_artifact_mtime:
                    ext_list.append(f"  - `{f}` (**NEW** \u2014 newer than latest artifact)")
                else:
                    ext_list.append(f"  - `{f}`")
            if ext_list:
                lines.append(f"- **External documents** ({len(ext_list)}) in `external/`:")
                lines.extend(ext_list)

        lines.append("")  # blank line between projects

    # Check for global policies
    policies_dir = os.path.join(projects_dir, "000-global", "policies")
    if os.path.isdir(policies_dir):
        policy_list = []
        for f in sorted(os.listdir(policies_dir)):
            fp = os.path.join(policies_dir, f)
            if os.path.isfile(fp):
                policy_list.append(f"  - `{f}`")
        if policy_list:
            lines.append("### Global Policies (000-global/policies/)")
            lines.extend(policy_list)
            lines.append("")

    context_text = "\n".join(lines)

    output = {
        "suppressOutput": True,
        "systemMessage": context_text,
    }
    print(json.dumps(output))


if __name__ == "__main__":
    main()

File: arckit-plugin/hooks/arckit-context.mjs
#!/usr/bin/env node
/**
 * ArcKit UserPromptSubmit Hook
 *
 * Pre-computes project context when any /arckit: command is run.
 * Injects project inventory, artifact lists, and external documents
 * via additionalContext so commands don't need to discover this themselves.
 *
 * Hook Type: UserPromptSubmit
 * Input (stdin): JSON with prompt, cwd, etc.
 * Output (stdout): JSON with additionalContext containing project context
 */

import { readdirSync } from 'node:fs';
import { join, resolve } from 'node:path';
import { DOC_TYPES, SUBDIR_MAP } from '../config/doc-types.mjs';
import {
  isDir, isFile, mtimeMs, readText,
  findRepoRoot, extractDocType, parseHookInput,
} from './hook-utils.mjs';

function docTypeName(code) {
  return DOC_TYPES[code]?.name || code;
}

// --- Main ---
const data = parseHookInput();

const userPrompt = data.prompt || '';

// Only run for /arckit: commands
if (!userPrompt.startsWith('/arckit:')) process.exit(0);

// Commands that don't need project context
const cmdMatch = userPrompt.match(/^\/arckit:([a-z_-]*)/);
if (cmdMatch) {
  const command = cmdMatch[1];
  if (['pages', 'customize', 'create', 'init', 'list', 'trello'].includes(command)) {
    process.exit(0);
  }
}

// Find repo root
const cwd = data.cwd || process.cwd();
const repoRoot = findRepoRoot(cwd);
if (!repoRoot) process.exit(0);

const projectsDir = join(repoRoot, 'projects');
if (!isDir(projectsDir)) process.exit(0);

// Read ArcKit version from plugin VERSION file
const pluginRoot = resolve(import.meta.url.replace('file://', ''), '..', '..');
const arckitVersion = readText(join(pluginRoot, 'VERSION'))?.trim() || 'unknown';

// Build context string
const lines = [];
lines.push('## ArcKit Project Context (auto-detected by hook)\n');
lines.push(`Repository: ${repoRoot}`);
lines.push(`ArcKit Version: ${arckitVersion}\n`);

// Count projects
const projectEntries = readdirSync(projectsDir)
  .filter(e => isDir(join(projectsDir, e)))
  .sort();
lines.push(`**${projectEntries.length} project(s) found:**\n`);

// Scan each project
for (const projectName of projectEntries) {
  const projectDir = join(projectsDir, projectName);

  // Extract project number
  let projectNumber = '';
  const pm = projectName.match(/^(\d{3})-/);
  if (pm) projectNumber = pm[1];

  lines.push(`### ${projectName}`);
  lines.push(`- **Path**: ${projectDir}`);
  if (projectNumber) lines.push(`- **Project ID**: ${projectNumber}`);

  // Scan for ARC-* artifacts in main project dir
  const artifactList = [];
  let artifactCount = 0;
  let newestArtifactMtime = 0;

  for (const f of readdirSync(projectDir).sort()) {
    const fp = join(projectDir, f);
    if (isFile(fp) && f.startsWith('ARC-') && f.endsWith('.md')) {
      const dtype = extractDocType(f) || f;
      const dname = docTypeName(dtype);
      artifactList.push(`  - \`${f}\` (${dname})`);
      artifactCount++;
      const amtime = mtimeMs(fp);
      if (amtime > newestArtifactMtime) newestArtifactMtime = amtime;
    }
  }

  // Also scan subdirectories (derived from SUBDIR_MAP + reviews)
  const subdirs = [...new Set(Object.values(SUBDIR_MAP)), 'reviews'];
  for (const subdir of subdirs) {
    const subPath = join(projectDir, subdir);
    if (isDir(subPath)) {
      for (const f of readdirSync(subPath).sort()) {
        const fp = join(subPath, f);
        if (isFile(fp) && f.startsWith('ARC-') && f.endsWith('.md')) {
          const dtype = extractDocType(f) || f;
          const dname = docTypeName(dtype);
          artifactList.push(`  - \`${subdir}/${f}\` (${dname})`);
          artifactCount++;
          const amtime = mtimeMs(fp);
          if (amtime > newestArtifactMtime) newestArtifactMtime = amtime;
        }
      }
    }
  }

  if (artifactCount > 0) {
    lines.push(`- **Artifacts** (${artifactCount}):`);
    lines.push(...artifactList);
  } else {
    lines.push('- **Artifacts**: none');
  }

  // Check for vendor directories and profiles
  const vendorsDir = join(projectDir, 'vendors');
  if (isDir(vendorsDir)) {
    const vendorDirs = [];
    const vendorProfiles = [];
    for (const vname of readdirSync(vendorsDir).sort()) {
      const vpath = join(vendorsDir, vname);
      if (isDir(vpath)) vendorDirs.push(`  - ${vname}`);
      else if (isFile(vpath) && vname.endsWith('-profile.md')) vendorProfiles.push(`  - ${vname}`);
    }
    if (vendorDirs.length > 0 || vendorProfiles.length > 0) {
      lines.push(`- **Vendors** (${vendorDirs.length + vendorProfiles.length}):`);
      lines.push(...vendorProfiles, ...vendorDirs);
    }
  }

  // Check for tech notes
  const techNotesDir = join(projectDir, 'tech-notes');
  if (isDir(techNotesDir)) {
    const noteList = [];
    for (const f of readdirSync(techNotesDir).sort()) {
      if (isFile(join(techNotesDir, f)) && f.endsWith('.md')) noteList.push(`  - ${f}`);
    }
    if (noteList.length > 0) {
      lines.push(`- **Tech Notes** (${noteList.length}):`);
      lines.push(...noteList);
    }
  }

  // Check for external documents
  const externalDir = join(projectDir, 'external');
  if (isDir(externalDir)) {
    const extList = [];
    for (const f of readdirSync(externalDir).sort()) {
      const fp = join(externalDir, f);
      if (!isFile(fp)) continue;
      if (f === 'README.md') continue;
      const extMtime = mtimeMs(fp);
      if (extMtime > newestArtifactMtime) {
        extList.push(`  - \`${f}\` (**NEW** \u2014 newer than latest artifact)`);
      } else {
        extList.push(`  - \`${f}\``);
      }
    }
    if (extList.length > 0) {
      lines.push(`- **External documents** (${extList.length}) in \`external/\`:`);
      lines.push(...extList);
    }
  }

  lines.push(''); // blank line between projects
}

// Check for global policies
const policiesDir = join(projectsDir, '000-global', 'policies');
if (isDir(policiesDir)) {
  const policyList = [];
  for (const f of readdirSync(policiesDir).sort()) {
    const fp = join(policiesDir, f);
    if (isFile(fp)) policyList.push(`  - \`${f}\``);
  }
  if (policyList.length > 0) {
    lines.push('### Global Policies (000-global/policies/)');
    lines.push(...policyList);
    lines.push('');
  }
}

const contextText = lines.join('\n');

const output = {
  hookSpecificOutput: {
    hookEventName: 'UserPromptSubmit',
    additionalContext: contextText,
  },
};
console.log(JSON.stringify(output));
